etl/assets.py

```diff
@@ -1,71 +1,96 @@
 
 
 @asset
-def wikipedia_articles_from_storage(
-    wikipedia_reader_resource: WikipediaReaderResource,
-):
+def wikipedia_articles_from_storage() -> RecordTuple:
     """Materialize an asset of Wikipedia articles."""
 
-    return tuple(wikipedia_reader_resource.read())
+    return RecordTuple(
+        records=tuple(
+            WikipediaReader(data_files_config=data_files_config_from_env_vars).read()
+        )
+    )
 
 
 @asset
 def wikipedia_articles_with_summaries(
-    wikipedia_articles_from_storage,
-    openai_generative_model_resource: OpenAiGenerativeModelResource,
-):
+    wikipedia_articles_from_storage: RecordTuple, config: OpenAiPipelineConfig
+) -> RecordTuple:
     """Materialize an asset of Wikipedia articles with summaries."""
 
-    enriched_wikipedia_articles = [
-        openai_generative_model_resource.enrich_record(wikipedia_article)
-        for wikipedia_article in wikipedia_articles_from_storage
-    ]
+    return RecordTuple(
+        records=tuple(
+            OpenAiGenerativeModelPipeline(config).enrich_record(wikipedia_article)
+            for wikipedia_article in wikipedia_articles_from_storage.records
+        )
+    )
+
+
+@asset
+def wikipedia_articles_with_summaries_json_file(
+    wikipedia_articles_with_summaries: RecordTuple,
+) -> None:
+    """Store the asset of Wikipedia articles with summaries as JSON."""
+
+    output_directory_path = output_config_from_env_vars.parse().directory_path
 
-    output_directory = Path(__file__).parent.absolute() / "data" / "output"
-    if not output_directory.exists():
-        os.mkdir(output_directory)
+    output_directory_path.mkdir(exist_ok=True)
 
-    enriched_wikipedia_output_directory = (
-        output_directory / "wikipedia_articles_with_summaries.txt"
+    enriched_wikipedia_output_file_path = (
+        output_directory_path / "wikipedia_articles_with_summaries.txt"
     )
 
-    with enriched_wikipedia_output_directory.open(
+    with enriched_wikipedia_output_file_path.open(
         mode="w"
     ) as enriched_wikipedia_output_file:
-        enriched_wikipedia_articles_as_json = [
-            json.dumps(enriched_wikipedia_article.model_dump(by_alias=True))
-            for enriched_wikipedia_article in enriched_wikipedia_articles
-        ]
 
-        enriched_wikipedia_output_file.writelines(enriched_wikipedia_articles_as_json)
+        enriched_wikipedia_output_file.writelines(
+            [
+                json.dumps(enriched_wikipedia_article.model_dump(by_alias=True))
+                for enriched_wikipedia_article in wikipedia_articles_with_summaries.records
+            ]
+        )
 
-    return enriched_wikipedia_articles
+
+@asset
+def documents_of_wikipedia_articles_with_summaries(
+    wikipedia_articles_with_summaries: RecordTuple, config: OpenAiPipelineConfig
+) -> DocumentTuple:
+    """Materialize an asset of Documents of Wikipedia articles with summaries."""
+
+    return DocumentTuple(
+        documents=create_documents(
```

Comment:
Nit: I'd just have `DocumentTuple.from_records` `@classmethod`. The utility function is awkwardly named.

Reply:
I am not sure if you meant I should remove `create_documents`, but I did that and placed it within `DocumentTuple.from_records`.

Reply:
Since `DocumentTuple.from_records` is the only method that requires that functionality.

Reply:
It's a factory function. I usually put factory functions as classmethods on the thing that's being created, unless I don't control that thing. (Unfortunately Python doesn't have extension methods like C#, Scala, and other languages.)